File: src/errata/models/registry.py
"""Build model adapters from config."""

from __future__ import annotations

from errata.config import settings
from errata.models.base import ModelAdapter


def get_adapter(model_id: str) -> ModelAdapter:
    """Return the appropriate adapter for a model ID."""
    if model_id.startswith("claude"):
        if not settings.anthropic_api_key:
            raise ValueError(f"ANTHROPIC_API_KEY not set (needed for {model_id})")
        from errata.models.anthropic import AnthropicAdapter
        return AnthropicAdapter(model_id, settings.anthropic_api_key)

    if model_id.startswith(("gpt-", "o1", "o3")):
        if not settings.openai_api_key:
            raise ValueError(f"OPENAI_API_KEY not set (needed for {model_id})")
        from errata.models.openai import OpenAIAdapter
        return OpenAIAdapter(model_id, settings.openai_api_key)

    if model_id.startswith("gemini"):
        if not settings.google_api_key:
            raise ValueError(f"GOOGLE_API_KEY not set (needed for {model_id})")
        from errata.models.gemini import GeminiAdapter
        return GeminiAdapter(model_id, settings.google_api_key)

    raise ValueError(f"Unknown model: {model_id!r}. Cannot determine provider.")


def list_adapters() -> list[ModelAdapter]:
    """Return adapters for all active models, skipping ones with missing keys."""
    adapters = []
    for model_id in settings.resolved_active_models:
        try:
            adapters.append(get_adapter(model_id))
        except ValueError as exc:
            # Warn but don't crash — display layer will show the warning
            import warnings
            warnings.warn(str(exc), stacklevel=2)
    return adapters

File: internal/models/registry.go
package models

import (
	"fmt"
	"strings"

	"github.com/suarezc/errata/internal/config"
)

// NewAdapter returns a ModelAdapter for the given model ID using cfg for API keys.
func NewAdapter(modelID string, cfg config.Config) (ModelAdapter, error) {
	switch {
	case strings.HasPrefix(modelID, "claude"):
		if cfg.AnthropicAPIKey == "" {
			return nil, fmt.Errorf("ANTHROPIC_API_KEY not set")
		}
		return NewAnthropicAdapter(modelID, cfg.AnthropicAPIKey), nil

	case strings.HasPrefix(modelID, "gpt-"),
		strings.HasPrefix(modelID, "o1"),
		strings.HasPrefix(modelID, "o3"):
		if cfg.OpenAIAPIKey == "" {
			return nil, fmt.Errorf("OPENAI_API_KEY not set")
		}
		return NewOpenAIAdapter(modelID, cfg.OpenAIAPIKey), nil

	case strings.HasPrefix(modelID, "gemini"):
		if cfg.GoogleAPIKey == "" {
			return nil, fmt.Errorf("GOOGLE_API_KEY not set")
		}
		return NewGeminiAdapter(modelID, cfg.GoogleAPIKey), nil

	default:
		return nil, fmt.Errorf("unknown model prefix for %q", modelID)
	}
}

// ListAdapters builds adapters for all resolved active models.
// Returns the adapter slice and a list of warning strings for skipped models.
func ListAdapters(cfg config.Config) ([]ModelAdapter, []string) {
	models := cfg.ResolvedActiveModels()
	var adapters []ModelAdapter
	var warnings []string

	for _, id := range models {
		a, err := NewAdapter(id, cfg)
		if err != nil {
			warnings = append(warnings, fmt.Sprintf("skipping %s: %v", id, err))
			continue
		}
		adapters = append(adapters, a)
	}
	return adapters, warnings
}
